### PR TITLE
Add support for symlinked magento root in errors pages / maintenance mode

### LINKED ISSUE #29618: [Issue] Add support for symlinked magento root in errors pages / maintenance mode
This issue is automatically created based on existing pull request: magento/magento2#26361: Add support for symlinked magento root in errors pages / maintenance mode

---------
If the magento root or magento/pub is symlinked into a diffrent directory the Error Processor (`Magento\Framework\Error\Processor`) does no longer generate a correct view file url.  
The `$this->_indexDir` variable is set to the path where it is symlinked and the `$this->_errorDir` variable is set to  the origin path. The str_replace in line 246 does no longer work. 

### Description (*)
If the magento root or the magento/pub directory is symlinked into a directory (for instance `/shop`) `$this->_indexDir` (the place where it is symlinked) is diffrent to `$this->_errorDir` (the "origin" path)

### Fixed Issues (if relevant)
None 

### Manual testing scenarios (*)

1. Symlink a magento into an existing document root 
    `ln -s /path/to/magento /path/to/document-root/`
2. `bin/magento maintenance:enable`

### Questions or comments
I know that this is not the most common use case. But it occured on one of my shops and the fix is relatively easy. If I can help, feel free to ask. Thanks!

### Contribution checklist (*)
 - [X] Pull request has a meaningful description of its purpose
 - [X] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by thampe**
If the magento root or magento/pub is symlinked into a diffrent directory the Error Processor (`Magento\Framework\Error\Processor`) does no longer generate a correct view file url.  
The `$this->_indexDir` variable is set to the path where it is symlinked and the `$this->_errorDir` variable is set to  the origin path. The str_replace in line 246 does no longer work. 

### Description (*)
If the magento root or the magento/pub directory is symlinked into a directory (for instance `/shop`) `$this->_indexDir` (the place where it is symlinked) is diffrent to `$this->_errorDir` (the "origin" path)

### Fixed Issues (if relevant)
None 

### Manual testing scenarios (*)

1. Symlink a magento into an existing document root 
    `ln -s /path/to/magento /path/to/document-root/`
2. `bin/magento maintenance:enable`

### Questions or comments
I know that this is not the most common use case. But it occured on one of my shops and the fix is relatively easy. If I can help, feel free to ask. Thanks!

### Contribution checklist (*)
 - [X] Pull request has a meaningful description of its purpose
 - [X] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#29618: Add support for symlinked magento root in errors pages / maintenance mode

### CODE REVIEW

File: pub/errors/processor.php [hunk 1]
```diff
@@ -257,10 +257,10 @@ public function getViewFileUrl()
         //The url needs to be updated base on Document root path.
         return $this->getBaseUrl() .
         str_replace(
-            str_replace('\\', '/', $this->_indexDir),
+            str_replace('\\', '/', realpath($this->_indexDir)),
```
Comment (lbajsarowicz) [inline pub/errors/processor.php]: 1. Do not use `realpath`, but `\Magento\Framework\Filesystem\DriverInterface::getRealPath($path)`
2. These actions should be performed in constructor, not getter.

**Review by lbajsarowicz — CHANGES_REQUESTED** [review]
Invalid use of `realpath()` + invalid place for such change.

**Review by lbajsarowicz — DISMISSED** [review]
Looks green for me : )

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7003](https://jira.corp.magento.com/browse/ENGCOM-7003) has been created to process this Pull Request
:eight_spoked_asterisk: @lbajsarowicz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by magento-engcom-team** [comment]
@thampe thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Hi @thampe.
During testing, we faced the issue.

**_Problem:_** If the magento root or magento/pub haven't any links, the wrong path is shown

**Manual testing scenario:**
1. Install Magento;
2. bin/magento maintenance:enable;
3. Go to  http://{magento url}/errors/default/
4. Open **View page source**

**Actual Result:**
There are extra slashes
screeen2

@thampe Could you take a look?

Thanks!

**Comment by thampe** [comment]
Hi, 

I will take a look at as soon as I can, sorry.

**Comment by thampe** [comment]
Hi @engcom-Alfa

could you take another look? I fixed the mentioned problem.

Thanks!

**Comment by marvinhinz** [comment]
@lbajsarowicz @engcom-Alfa ? :)

**Comment by sidolov** [comment]
@magento create issue

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @thampe,
Seems like this issue appears just because of incorrect web server configuration.

Could you explain what it the use case when magento root or magento/pub should be symlinked into a different directory the Error Processor?

Could you maybe add additional info with some screenshot of correct and incorrect behavior?

**Comment by ihor-sviziev** [comment]
@thampe could you reply to https://github.com/magento/magento2/pull/26361#pullrequestreview-491272902 ?

**Comment by sidolov** [comment]
@thampe  I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Comment by thampe** [comment]
Hi,

all right one last try :).

>Seems like this issue appears just because of incorrect web server configuration.

As I stated above there is a case where you install the magento software in a diffrent place and it needs to run from a subdirectory. For instance you have your CMS in the document root and a magento accessible from `/shop`. Of course you can do with an nginx or apache configuration, but you could also just merge this tiny PR and make _your_ software better :).

Sure we can discuss another nine months about the usefulness of this PR. Honestly I do no longer see the point in it. You can now either merge it or leave it. I do not need this bureaucracy in my life.

Do not get me wrong I think magento and the people who are building it are amazing but this process needs to some improvements.

Greetings
\- Thomas

**Comment by ihor-sviziev** [comment]
Hi @thampe,

Thank you so much for your reply! I do agree with you that such configuration could be useful in some cases. Unfortunately there are so many cases used by different people and supporting all of these cases sometimes doesn't really make sense.

From what you said above - it really make sense to accept your PR. I believe this might be quite frequent case that people just fixing themselves and not reporting it to Magento.

I'm reopening this PR.

> Do not get me wrong I think magento and the people who are building it are amazing but **this process needs to some improvements.**

Sure, you're totally right! We already working on improving it by reviewing older PRs and doing few other improvements.

File: pub/errors/processor.php [hunk 2]
```diff
@@ -440,7 +466,7 @@ protected function _renderPage($template)
     protected function _getFilePath($file, $directories = null)
     {
         if ($directories === null) {
-            $directories[] = $this->_errorDir;
+            $directories[] = $this->errorDirectory;
```
Comment (ihor-sviziev) [inline pub/errors/processor.php]:
```suggestion
            $directories[] = $this->errorDir;
```

File: pub/errors/processor.php [hunk 3]
```diff
@@ -458,10 +484,10 @@ protected function _getFilePath($file, $directories = null)
      */
     protected function _getTemplatePath($template)
     {
-        $directories[] = $this->_errorDir . $this->_config->skin . '/';
+        $directories[] = $this->errorDirectory . $this->_config->skin . '/';
```
Comment (ihor-sviziev) [inline pub/errors/processor.php]:
```suggestion
        $directories[] = $this->errorDir . $this->_config->skin . '/';
```

File: pub/errors/processor.php [hunk 4]
```diff
@@ -458,10 +484,10 @@ protected function _getFilePath($file, $directories = null)
...
 
         if ($this->_config->skin != self::DEFAULT_SKIN) {
-            $directories[] = $this->_errorDir . self::DEFAULT_SKIN . '/';
+            $directories[] = $this->errorDirectory . self::DEFAULT_SKIN . '/';
```
Comment (ihor-sviziev) [inline pub/errors/processor.php]:
```suggestion
            $directories[] = $this->errorDir . self::DEFAULT_SKIN . '/';
```

File: pub/errors/processor.php [hunk 5]
```diff
@@ -718,7 +744,7 @@ protected function _validate()
      */
     protected function _setSkin($value, \stdClass $config = null)
     {
-        if (preg_match('/^[a-z0-9_]+$/i', $value) && is_dir($this->_errorDir . $value)) {
+        if (preg_match('/^[a-z0-9_]+$/i', $value) && is_dir($this->errorDirectory . $value)) {
```
Comment (ihor-sviziev) [inline pub/errors/processor.php]:
```suggestion
        if (preg_match('/^[a-z0-9_]+$/i', $value) && is_dir($this->errorDir . $value)) {
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @thampe,
In general your PR looks good. The only issue - we have notices due to different variable names in constructor and usage in the methods and because of that we have notices in integration tests:
image

Could you do following things? 
- apply suggested changes, they should fix the notices
- merge the latest changes from 2.4-develop branch

**Comment by engcom-Charlie** [comment]
@thampe please look at the requested changes.
Thank you.

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by thampe** [comment]
Hi,

I merge the current `2.4-develop` branch and commited the suggested chnages.

Thanks again!

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8325](https://jira.corp.magento.com/browse/ENGCOM-8325) has been created to process this Pull Request

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Hi @thampe .
Unfortunately, during the testing, we faced the issue.

**Manual testing scenario:**

1. Install Magento ( `default` mode);
2. Enable maintenance mode ( `php bin/magento maintenance:enable`);
3. Go to Storefront ( **Home Page** for ex.);
4. Open **exception.log** file from `var/log` directory;

**Actual Result:** :heavy_multiplication_x: Undefined property.... $errorDirectory

Screenshot from 2020-10-12 10-21-09

@thampe Could you take a look?
Thanks!

**Comment by engcom-Charlie** [comment]
Hi @thampe can you please look at https://github.com/magento/magento2/pull/26361#pullrequestreview-506360011? Otherwise, we can't proceed with your PR.
Thank you.

**Comment by thampe** [comment]
Hi,

yes, I will take a look at this, at the moment I am relatively busy, but I will try to do it this week!

Sorry!

File: pub/errors/processor.php [hunk 6]
```diff
@@ -254,13 +279,14 @@ public function processReport()
      */
     public function getViewFileUrl()
     {
+        $indexDirectory = str_replace('\\', '/', $this->indexDir);
+        $errorDirectory = str_replace('\\', '/', $this->errorDirectory);
```
Comment (ihor-sviziev) [inline pub/errors/processor.php:283]:
```suggestion
        $errorDirectory = str_replace('\\', '/', $this->errorDir);
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Could you apply following change?

**Comment by engcom-Charlie** [comment]
Hi @thampe please look at https://github.com/magento/magento2/pull/26361#pullrequestreview-506360011 https://github.com/magento/magento2/pull/26361#discussion_r512548051. Otherwise, we can't proceed with your PR.
Thank you.

**Comment by ihor-sviziev** [comment]
Hi @thampe,
FYI there were some changes in recent release and they were fixing the same method. Could you make sure the issue still actual. If so - please resolve conflicts

**Comment by engcom-Charlie** [comment]
Hi @thampe, could you please look at https://github.com/magento/magento2/pull/26361#issuecomment-718636768?
Thank you.

**Comment by sidolov** [comment]
@thampe  I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Pull request closed without merging**